Add Versioning Guide to CONTRIBUTING file

rye opened the pull request:
This is a fairly straightforward implementation of the early spec reached in #57, which is still awaiting feedback from @samontea.

Please do provide feedback as you see fit.

rye committed:
Add Branching/Versioning Scheme section to CONTRIBUTING

This is a rough approximation of the result from the discussion in
https://github.com/kotct/dot/issues/57.

rye committed:
Merge branch 'revamp-contributing-file' into versioning-guide

cg505 commented:
I thought we decided to ditch the `release/` branches and just tag instead?

rye commented:
From [this comment](https://github.com/kotct/dot/issues/57#issuecomment-315226276) it sounded like 

>the `release-` branch can get discarded after the release is released and the branch is in `master`

Basically we'd keep them around for just a short time while bumping versions and potentially adding to changelogs

rye commented:
But we would still use tags, and the `release-` branches would get deleted after the release took place.

cg505 commented on CONTRIBUTING.md:
```diff
@@ -31,3 +31,32 @@ In general, here's a suggested workflow:
+- `feature/` can be used as a prefix for feature branches, but this is not recommended.
+  (Just use `kebab-case` for your branch names, we assume that anything without a special prefix is a normal feature PR.)
+- `hotfix/` should be used as a prefix for hotfix branches.
+  If a hotfix targets a release, (typical) a `release-vX.Y.Z` branch should be used as the base of the PR instead; submit a PR for both the `release-vX.Y.Z` and the `hotfix/` branch.
```
typo: "typically"

cg505 commented on CONTRIBUTING.md:
```diff
@@ -31,3 +31,32 @@ In general, here's a suggested workflow:
+
+In addition, we use the following scheme for additional branches:
+
+- `release-vX.Y` and `release-vX.Y.Z` branches are branches used to control versioning.
```
What is the point of the `release-vX.Y` branch? Seems unnecessary to me. The one possible use is leaving one around that always points to the latest release in that minor version, as a convenience. This mostly seems unnecessary though.

rye replied:
Bumping a version is something that should probably not really be done by people submitting changes.  That's project-level work that should really just be done by the maintainers when they know what they want to merge.  We'd delete the branch, but people would be able to create the branch again if we needed to add something to a previous version in a patch release.

cg505 replied:
So I understand how you would use a `release-v2.1.0` branch, for instance, but what's the point of the `release-v2.1` branch?

Also, can we consider doing `release/` so that we don't have two different types of prefixes?

rye replied:
Further discussion should probably take place on #57.

cg505 commented on CONTRIBUTING.md:
```diff
@@ -31,3 +31,32 @@ In general, here's a suggested workflow:
+    2. Create your `release-v` branch.
+    3. Bump the version on the `release-v` branch, either through a PR on a separate branch or something like that.
+       If you have any release notes to add, do so here.
+    4. Merge the branch back into `master` via a PR, including writing release notes and such.
```
Imo the PR should target `stable`, and the tag is on that merge commit into `stable`. `stable` then can be merged back into master. It doesn't make sense to have active devel going on in master, make a release branch, and then merge that right back into `master`. The diff should be 0.

rye replied:
I would agree if you're saying that any development work off of `stable` should take place and be directed back to the `stable` branch, but some people like @cg505 and myself would probably be working off of `master`.  Again, making a `release-vX.Y` branch is intended for work _related to the release process_, (incl. merging PR's) and nothing else.  The point of branching off of one of these branches is to have a nice place to merge back into, but I don't expect 3rd parties (or even contributors who are submitting their own changes) to properly take care of documenting their changes.

I don't understand what you mean by
>The diff should be 0.

cg505 replied:
Popping this back into the issue (#57) for more discussion.

cg505 reviewed (changes requested):
I disagree with how the release process is put forward. Discuss as needed, but if substantial discussion must be had, move it to the issue instead of having it in the PR.

rye committed:
Adjust ambiguous parenthesized "typical" a bit

rye committed:
Make release branch naming more consistent

This is more consistent with the guidelines for other schemes set
forth in this document.

cg505 commented on CONTRIBUTING.md:
```diff
@@ -31,3 +31,32 @@ In general, here's a suggested workflow:
+    2. Create your `release/vX.Y(.Z)` branch.
+    3. Bump the version on the `release/vX.Y(.Z)` branch, either through a PR on a separate branch or something like that.
+       If you have any release notes to add, do so here.
+    4. Merge the branch back into `master` via a PR, including writing release notes and such.
```
We decided to merge into `stable` via a PR?

cg505 replied:
Also we should probably get at least one approval (on the release notes and on the release) in order to do this merge. Should already be enforced with our current PR rules. (Don't worry about changing the actual text of the file here, this was just meant to be a note.)

rye replied:
Note that I haven't updated this file at all since our discussion in #57, I was just looking to highlight the actual issues in this file.  Will fix shortly.

cg505 commented on CONTRIBUTING.md:
```diff
@@ -31,3 +31,32 @@ In general, here's a suggested workflow:
+    3. Bump the version on the `release/vX.Y(.Z)` branch, either through a PR on a separate branch or something like that.
+       If you have any release notes to add, do so here.
+    4. Merge the branch back into `master` via a PR, including writing release notes and such.
+    5. Create a release targeting the PR merge commit.
```
Add a step to merge the release back into `master`. No PR needed for this one, just do it.

rye replied:
Isn't `master` merge-locked?

rye replied:
This means we have to have a reviewed PR before we can merge.  That's why I was thinking about doing it the other way; merge to `master` then merge to `stable`.

cg505 replied:
mmmm. Since an admin will be doing the release, we can just open a PR, then override checks and immediately merge I guess.

rye replied:
Yeah, it'd work that way.

cg505 commented on CONTRIBUTING.md:
```diff
@@ -31,3 +31,32 @@ In general, here's a suggested workflow:
+- `feature/` can be used as a prefix for feature branches, but this is not recommended.
+  (Just use `kebab-case` for your branch names, we assume that anything without a special prefix is a normal feature PR.)
+- `hotfix/` should be used as a prefix for hotfix branches.
+  If a hotfix targets a release, (which is the typical case) a `release/vX.Y.Z` branch should be used as the base of the PR instead; submit a PR for both the `release/vX.Y.Z` and the `hotfix/` branch.
```
>submit a PR for both the `release/vX.Y.Z` and the `hotfix/` branch.

What does it mean to submit a PR for the `hotfix/` branch? Aren't we working in the `hotfix/` branch? We PR `hotfix/*` into `release/vX.Y.Z` and then (if X.Y is the current minor version) PR `release/vX.Y.Z` into stable and tag. Right?

rye replied:
Yeah, that's right.  `hotfix` goes into `release`, `release` goes into `stable`, `stable` gets tagged, `stable` gets merged back to `master`.

cg505 replied:
Okay. Maybe clarify this language to say what is being merged into what in those PRs.

rye replied:
Yep.

cg505 reviewed (changes requested):
Looking pretty good. A few small remaining issues.

rye committed:
Capitalize step 1 in release process

rye committed:
Make `master` release step 3 more verbose

rye committed:
Make release steps 4 and 5 more accurate to our decision

rye committed:
Add release step to merge back to master

rye committed:
Make hotfix procedures a bit more clear

cg505 reviewed (approved):
feel free to merge

rye commented:
Whee! Merging now.
